# **Capstone Project - The Battle of Neighborhoods** #

## **1. Introduction: Business Problem** ##

In this project, our main objective is to look for an optimal location to open a Chinese restaurant. Main target audiences are aspiring chefs or business owners who are looking to open a Chinese restaurant in the city of Toronto.

One of the key considerations when deciding to open a restaurant is to firstly identify a location which is not overly crowded with existing restaurants as it means that competition will be greater. Specifically, the ideal location should have minimal or close to zero existing Chinese restaurants to minimize competition. Besides competition, the ideal location should also be in an area which is bustling with a significant population of Chinese race to cater to their palettes.

We will make use of data analytics to create a model that aims to recommend the ideal location which addresses the 3 key criteria identified above: (1) number of existing restaurants (2) number of existing Chinese restaurants (3) population with high number of Chinese. The final recommendations will be tabled together to allow targeted stakeholders to weigh out the pros and cons before deciding on the final location. 


## **2. Data** ##

- Neighborhood data for the city of Toronto. For this, there is a Wikipedia page that contains all the information that we need to explore and cluster the neighborhoods in Toronto
https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

- Data of number of existing restaurants and type of restaurants in each neighborhood to be obtained from Foursquare API
- Coordinates of each neighborhood: http://cocl.us/Geospatial_data
- Demographics of each neighborhood: https://open.toronto.ca/dataset/wellbeing-toronto-demographics/

### **2.1. Data Cleansing** ###

In [1]:
#download all dependencies and packages required

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')



Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-1.22.0-pyh9f0ad1d_0



geopy-1.22.0         | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: 

In [2]:
df_demo = pd.read_excel('wellbeing-toronto-demographics.xlsx')
new_header = df_demo.iloc[0]
df_demo = df_demo[1:]
df_demo.columns = new_header
df_demo.rename(columns={'Neighbourhood': 'Neighborhood'}, inplace =True)
df_demo.head()

0     Neighborhood    Chinese    South Asian    Black    Filipino  \
1         Milliken      16790           3780     1365        1145   
2          Steeles      16705           1895      660         755   
3  Agincourt North      16565           5160     1530        1355   
4       L'Amoreaux      16455           8285     3875        1905   
5  Willowdale East      14860           1700      845         520   

0    Latin American    Southeast Asian    Arab    West Asian    Korean  \
1               125                290     170           135        20   
2                50                 95     360           115       140   
3               230                230     370            75       155   
4               385                635     985           930       205   
5               440                190     485          3395      4265   

0    Japanese  
1          60  
2          55  
3         135  
4         245  
5         285

In [3]:
#Use OpenCage geocoder to obtain coordinates of neighborhoods in df_demo
!pip install opencage
from opencage.geocoder import OpenCageGeocode

In [4]:
#Obtain coordinates of each street in df_rent and put into df_demo
key = '9660749f28fa47d5a2877208870f5a68' #OpenCage API key
geocoder = OpenCageGeocode(key)

list_lat = []   # create empty lists

list_lon = []
for index, row in df_demo.iterrows(): # iterate over rows in dataframe

    Neighborhood = row['Neighborhood']       
    query = str(Neighborhood)+', Toronto'

    results = geocoder.geocode(query)   
    lat = results[0]['geometry']['lat']
    lng = results[0]['geometry']['lng']

    list_lat.append(lat)
    list_lon.append(lng)

# Create new columns from lists    

df_demo['Latitude'] = list_lat   

df_demo['Longitude'] = list_lon

df_demo.head()

0     Neighborhood    Chinese    South Asian    Black    Filipino  \
1         Milliken      16790           3780     1365        1145   
2          Steeles      16705           1895      660         755   
3  Agincourt North      16565           5160     1530        1355   
4       L'Amoreaux      16455           8285     3875        1905   
5  Willowdale East      14860           1700      845         520   

0    Latin American    Southeast Asian    Arab    West Asian    Korean  \
1               125                290     170           135        20   
2                50                 95     360           115       140   
3               230                230     370            75       155   
4               385                635     985           930       205   
5               440                190     485          3395      4265   

0    Japanese   Latitude  Longitude  
1          60  43.823174 -79.301763  
2          55  43.816178 -79.314538  
3         135  43.808038 -79.266439  
4         245  43.799003 -79.305967  
5         285  43.761510 -79.410923

In [5]:
df_demo.shape

(140, 13)

### **2.2. Explore and Cluster Neighborhoods in Toronto** ###

In [6]:
#use geopy library to get latitude and longitude of Toronto
address = 'Toronto, ON'

geolocator = Nominatim(user_agent = 'toronto_explorer')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinates of Toronto are 43.6534817, -79.3839347.


In [7]:
#create a map of Toronto
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)
map_toronto

In [8]:
#add markers to the map which superimposes the neighborhoods on top
for lat, lng, neighborhood in zip(df_demo['Latitude'], df_demo['Longitude'], df_demo['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
    [lat, lng],
    radius = 5,
    popup = label,
    color = 'blue',
    fill = 'True',
    fill_color = '#3186cc',
    fill_opacity = 0.7,
    parse_html = False).add_to(map_toronto)

map_toronto

In [9]:
#utilise Foursquare API to explore neighborhoods and segment them
#define Foursquare credentials and version

CLIENT_ID = 'XJXUFZ3OOBCG21K2A4PXY5UWVACMEUEQVFWKPIBZX2MSG1UU' 
CLIENT_SECRET = 'WGYPJNQY40GGKHX5DCOXWICCV3RALMM44MXKMJAARCL0RUHR'
VERSION = '20180605' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XJXUFZ3OOBCG21K2A4PXY5UWVACMEUEQVFWKPIBZX2MSG1UU
CLIENT_SECRET:WGYPJNQY40GGKHX5DCOXWICCV3RALMM44MXKMJAARCL0RUHR


In [10]:
#Explore First Neighborhood in Data Frame
df_demo.loc[1, 'Neighborhood']

'Milliken'

In [11]:
#get Milliken's latitude and longitude values
Milliken_latitude = df_demo.loc[1, 'Latitude']
Milliken_longitude = df_demo.loc[1, 'Longitude']
Milliken_name = df_demo.loc[1, 'Neighborhood']

print('Latitude and longitude of {} are {} and {}.'.format(Milliken_name, Milliken_latitude, Milliken_longitude))

Latitude and longitude of Milliken are 43.8231743 and -79.3017626.


In [12]:
#get top 100 venues that are within radius of 500m from Milliken

LIMIT = 100
radius = 500

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    Milliken_latitude, 
    Milliken_longitude, 
    radius, 
    LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=XJXUFZ3OOBCG21K2A4PXY5UWVACMEUEQVFWKPIBZX2MSG1UU&client_secret=WGYPJNQY40GGKHX5DCOXWICCV3RALMM44MXKMJAARCL0RUHR&v=20180605&ll=43.8231743,-79.3017626&radius=500&limit=100'

In [13]:
#using get request to obtain the results
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ed50ba2c8cff2001b0efa6c'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Scarborough',
  'headerFullLocation': 'Scarborough',
  'headerLocationGranularity': 'city',
  'totalResults': 35,
  'suggestedBounds': {'ne': {'lat': 43.8276743045, 'lng': -79.29553706206312},
   'sw': {'lat': 43.818674295499996, 'lng': -79.30798813793689}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '56945d1c498e11466e96405f',
       'name': 'Planet Fitness North Scarborough',
       'location': {'lat': 43.824095167666584,
        'lng': -79.30141064389495,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.824095167666584,
          'lng': -79.3014

In [14]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
    
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [15]:
#clean the json and structure it into a pandas dataframe
venues = results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name              categories        lat  \
0       Planet Fitness North Scarborough                     Gym  43.824095   
1          Deer Garden Signatures 鹿園魚湯米線            Noodle House  43.821898   
2                         Nichiban Sushi        Sushi Restaurant  43.823172   
3                        Aka-Oni Izakaya     Japanese Restaurant  43.822372   
4                      Sun's Kitchen 拉麵王      Chinese Restaurant  43.825282   
5                    Allan's Pastry Shop                  Bakery  43.820953   
6      Uncle Tetsu's Japanese Cheesecake                  Bakery  43.825150   
7          New Northern Dumplings 新北方餃子館     Dumpling Restaurant  43.821886   
8     Kim Po Vietnamese Cuisine - 金寶越南美食   Vietnamese Restaurant  43.823292   
9                 Fish Ball Place 真之味小食屋             Snack Place  43.825290   
10                           Tim Hortons             Coffee Shop  43.825423   
11           Pacific Heritage Town 太古民族村      Miscellaneous Shop  43.825796   
12                        RBC Royal Bank                    Bank  43.825195   
13                  Brothers Bakery 兄弟餅店                  Bakery  43.825443   
14       Tak Fu Seafood Restaurant 德福點心皇      Chinese Restaurant  43.822633   
15                        Fat Ninja Bite     Japanese Restaurant  43.823043   
16              Ga Bin Korean Restaurant       Korean Restaurant  43.822475   
17                          Petro-Canada             Gas Station  43.823847   
18    東東蛋卷皇 Tung Tung Dan Quen King Ltd.        Asian Restaurant  43.826140   
19            Ming's Noodle Cafe 明記粥麵茶餐廳    Hong Kong Restaurant  43.820940   
20            Real Fruit Bubble Tea 真果茶坊         Bubble Tea Shop  43.825483   
21                  Papa Chang's Express    Taiwanese Restaurant  43.825612   
22                               Wendy’s    Fast Food Restaurant  43.825438   
23                Lychee Bay Cuisine 荔枝灣      Chinese Restaurant  43.825232   
24  Verdant Garden Chinese Restaurant 翠苑    Cantonese Restaurant  43.822834   
25              Kyushou Japanese Noodles     Japanese Restaurant  43.824985   
26           Hunan Style Restaurant 湘村發現      Chinese Restaurant  43.822442   
27            Tian Fu Chinese Restaurant      Chinese Restaurant  43.825418   
28                      Aqua Inspiration               Pet Store  43.821486   
29                          Club Kennedy       Food & Drink Shop  43.821916   
30               Japanese Express 櫻花日本料理     Japanese Restaurant  43.825611   
31                     Pacific Mall 太古廣場           Shopping Mall  43.826152   
32                Neohome Houseware 千色家品  Furniture / Home Store  43.826928   
33                 Prince B.B.Q. 皇子燒臘美食店      Chinese Restaurant  43.826781   
34                  Watami Sushi & Grill     Japanese Restaurant  43.826758   

          lng  
0  -79.301411  
1  -79.298857  
2  -79.306064  
3  -79.298905  
4  -79.306231  
5  -79.304564  
6  -79.305954  
7  -79.298751  
8  -79.305257  
9  -79.306202  
10 -79.297787  
11 -79.306242  
12 -79.299575  
13 -79.306984  
14 -79.298958  
15 -79.306446  
16 -79.299044  
17 -79.306098  
18 -79.304602  
19 -79.305356  
20 -79.305847  
21 -79.306341  
22 -79.297706  
23 -79.298709  
24 -79.300832  
25 -79.300537  
26 -79.299049  
27 -79.298917  
28 -79.305499  
29 -79.305883  
30 -79.306097  
31 -79.306312  
32 -79.304811  
33 -79.305130  
34 -79.305247

In [20]:
#Explore All Neighborhoods in Toronto

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
toronto_venues = getNearbyVenues(names = df_demo['Neighborhood'],
                                   latitudes = df_demo['Latitude'],
                                   longitudes = df_demo['Longitude']
                                  )

Milliken
Steeles
Agincourt North
L'Amoreaux
Willowdale East
Agincourt South-Malvern West
Tam O'Shanter-Sullivan
Hillcrest Village
South Riverdale
Don Valley Village
Kensington-Chinatown
Pleasant View
Bayview Village
Newtonbrook East
Woburn
Bayview Woods-Steeles
Malvern
Trinity-Bellwoods
Banbury-Don Mills
Dorset Park
Bendale
Parkwoods-Donalda
York University Heights
Bay Street Corridor
St.Andrew-Windfields
Greenwood-Coxwell
Dovercourt-Wallace Emerson-Juncti
Waterfront Communities-The Island
Rouge
Church-Yonge Corridor
Henry Farm
Glenfield-Jane Heights
Flemingdon Park
North Riverdale
Clairlea-Birchmount
Annex
Willowdale West
Regent Park
Wexford/Maryvale
Danforth-East York
Little Portugal
Kennedy Park
Ionview
Palmerston-Little Italy
Dufferin Grove
Taylor-Massey
North St.James Town
Newtonbrook West
Niagara
South Parkdale
Black Creek
East End-Danforth
Eglinton East
Birchcliffe-Cliffside
Blake-Jones
Islington-City Centre West
Lansing-Westgate
Oakridge
Woodbine-Lumsden
Roncesvalles
Mount Plea

In [22]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Milliken              43.823174              -79.301763   
1     Milliken              43.823174              -79.301763   
2     Milliken              43.823174              -79.301763   
3     Milliken              43.823174              -79.301763   
4     Milliken              43.823174              -79.301763   

                              Venue  Venue Latitude  Venue Longitude  \
0  Planet Fitness North Scarborough       43.824095       -79.301411   
1     Deer Garden Signatures 鹿園魚湯米線       43.821898       -79.298857   
2                    Nichiban Sushi       43.823172       -79.306064   
3                   Aka-Oni Izakaya       43.822372       -79.298905   
4                 Sun's Kitchen 拉麵王       43.825282       -79.306231   

        Venue Category  
0                  Gym  
1         Noodle House  
2     Sushi Restaurant  
3  Japanese Restaurant  
4   Chinese Restaurant

In [23]:
#check how many venues were returned for each neighborhood
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                               
Agincourt North                                       25   
Agincourt South-Malvern West                           3   
Alderwood                                              8   
Annex                                                 47   
Banbury-Don Mills                                      4   
Bathurst Manor                                         4   
Bay Street Corridor                                   42   
Bayview Village                                       14   
Bayview Woods-Steeles                                  3   
Bedford Park-Nortown                                  15   
Beechborough-Greenbrook                               15   
Bendale                                                8   
Birchcliffe-Cliffside                                 15   
Black Creek                                            4   
Blake-Jones                                           15   
Briar Hill-Belgravia                                  15   
Bridle Path-Sunnybrook-York Mills                     15   
Broadview North                                        9   
Brookhaven-Amesbury                                   15   
Cabbagetown-South St.James Town                       15   
Caledonia-Fairbank                                    12   
Casa Loma                                             22   
Centennial Scarborough                                 1   
Church-Yonge Corridor                                 35   
Clairlea-Birchmount                                   15   
Clanton Park                                           4   
Cliffcrest                                             7   
Corso Italia-Davenport                                23   
Danforth                                              30   
Danforth-East York                                    30   
Don Valley Village                                    11   
Dorset Park                                           10   
Dovercourt-Wallace Emerson-Juncti                     15   
Downsview-Roding-CFB                                  15   
Dufferin Grove                                        57   
East End-Danforth                                     15   
Edenbridge-Humber Valley                               4   
Eglinton East                                          6   
Elms-Old Rexdale                                       5   
Englemount-Lawrence                                    6   
Eringate-Centennial-West Deane                        15   
Etobicoke West Mall                                   13   
Flemingdon Park                                       14   
Forest Hill North                                      4   
Forest Hill South                                      4   
Glenfield-Jane Heights                                21   
Greenwood-Coxwell                                     15   
Guildwood                                              4   
Henry Farm                                             3   
High Park North                                        7   
High Park-Swansea                                     15   
Highland Creek                                         2   
Hillcrest Village                                     13   
Humber Heights-Westmount                               4   
Humber Summit                                          5   
Humbermede                                            15   
Humewood-Cedarvale                                    15   
Ionview                                                4   
Islington-City Centre West                             4   
Junction Area                                         54   
Keelesdale-Eglinton West                               5   
Kennedy Park                                           5   
Kensington-Chinatown                                  53   
Kingsview Village-The Westway                         15   
Kingsway South                                        33   
L'Amoreaux     

In [24]:
#find out number of unique categories
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 261 unique categories.


### **2.3. Analyze each Neighborhood in Toronto** ###

In [25]:
#onehot coding
toronto_venues_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix='', prefix_sep='')

#add neighborhood, latitude and longitude column into dataframe
toronto_venues_onehot['Neighborhood'] = toronto_venues['Neighborhood']
toronto_venues_onehot['Latitude'] = toronto_venues['Venue Latitude']
toronto_venues_onehot['Longitude'] = toronto_venues['Venue Longitude']


#move neighborhood column into first column
fixed_columns = [toronto_venues_onehot.columns[-1]] + list(toronto_venues_onehot.columns[:-1])
toronto_venues_onehot = toronto_venues_onehot[fixed_columns]

toronto_venues_onehot.head()

Longitude  ATM  Accessories Store  Afghan Restaurant  American Restaurant  \
0 -79.301411    0                  0                  0                    0   
1 -79.298857    0                  0                  0                    0   
2 -79.306064    0                  0                  0                    0   
3 -79.298905    0                  0                  0                    0   
4 -79.306231    0                  0                  0                    0   

   Animal Shelter  Antique Shop  Arcade  Art Gallery  Art Museum  \
0               0             0       0            0           0   
1               0             0       0            0           0   
2               0             0       0            0           0   
3               0             0       0            0           0   
4               0             0       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   Automotive Shop  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0                0          0           0       0     0    0               0   
1                0          0           0       0     0    0               0   
2                0          0           0       0     0    0               0   
3                0          0           0       0     0    0               0   
4                0          0           0       0     0    0               0   

   Basketball Court  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0      0                0         0           0   
1                 0      0                0         0           0   
2                 0      0                0         0           0   
3                 0      0                0         0           0   
4                 0      0                0         0           0   

   Belgian Restaurant  Big Box Store  Bistro  Bookstore  Botanical Garden  \
0                   0              0       0          0                 0   
1                   0              0       0          0                 0   
2                   0              0       0          0                 0   
3                   0              0       0          0                 0   
4                   0              0       0          0                 0   

   Boutique  Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  \
0         0              0                     0               0        0   
1         0              0                     0               0        0   
2         0              0                     0               0        0   
3         0              0                     0               0        0   
4         0              0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0            0                0         0             0              0   
1            0                0         0             0              0   
2            0                0         0             0              0   
3            0                0         0             0              0   
4            0                0         0             0              0   

   Bus Line  Bus Stop  Butcher  Café  Cajun / Creole Restaurant  Camera Store  \
0         0         0        0     0                          0             0   
1         0         0        0     0                          0             0   
2         0         0        0     0                          0             0   
3         0         0        0     

In [26]:
#group by neighborhood and find the mean of frequency of occurence for each category
toronto_grouped = toronto_venues_onehot.groupby('Neighborhood').sum().reset_index()
toronto_grouped.head()

Neighborhood    Longitude  ATM  Accessories Store  \
0               Agincourt North -1981.714451    0                  0   
1  Agincourt South-Malvern West  -237.777995    0                  0   
2                     Alderwood  -636.360463    0                  0   
3                         Annex -3732.132140    0                  0   
4             Banbury-Don Mills  -317.427164    0                  0   

   Afghan Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0                  0                    0               0             0   
1                  0                    0               0             0   
2                  0                    0               0             0   
3                  0                    0               0             0   
4                  0                    0               0             0   

   Arcade  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0            0           0                    0                 0   
1       0            0           0                    0                 0   
2       0            0           0                    0                 0   
3       0            0           0                    0                 0   
4       0            0           0                    0                 0   

   Athletics & Sports  Auto Dealership  Automotive Shop  BBQ Joint  \
0                   0                0                0          0   
1                   0                0                0          0   
2                   0                0                0          0   
3                   0                0                0          0   
4                   0                0                0          0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Basketball Court  Beach  \
0           0       2     2    0               0                 0      0   
1           0       0     0    0               0                 0      0   
2           0       0     0    0               0                 0      0   
3           0       1     0    0               0                 0      0   
4           0       0     0    0               0                 0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Big Box Store  \
0                0         0           1                   0              0   
1                0         0           0                   0              0   
2                0         0           0                   0              0   
3                0         0           0                   0              0   
4                0         0           0                   0              0   

   Bistro  Bookstore  Botanical Garden  Boutique  Bowling Alley  \
0       0          0                 0         0              0   
1       0          0                 0         0              0   
2       0          0                 0         0              0   
3       2          1                 0         0              0   
4       0          0                 1         0              0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               1        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Bus Stop  \
0                0         0             0              0         0         0   
1                0         0             0              0         0         0   
2                0         0             0              0         0         0   
3                0         0             0              1         0         0   
4                0         0             0              0         0         0   

   Butcher  Café  Cajun / Creole

In [27]:
df_chinese = toronto_grouped[['Neighborhood', 'Chinese Restaurant']]
df_chinese.sort_values(by=['Chinese Restaurant'], ascending = False).head()

Neighborhood  Chinese Restaurant
76          Milliken                   6
0    Agincourt North                   2
65        L'Amoreaux                   2
88   North Riverdale                   2
110  South Riverdale                   2

In [28]:
#print each neighborhood with its top 5 venues
num_top_venues = 5
toronto_grouped.drop('Latitude', axis=1, inplace=True)

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt North----
                venue  freq
0              Bakery   2.0
1  Chinese Restaurant   2.0
2                Bank   2.0
3      Ice Cream Shop   1.0
4      Discount Store   1.0


----Agincourt South-Malvern West----
                       venue  freq
0             Breakfast Spot   1.0
1                     Lounge   1.0
2  Latin American Restaurant   1.0
3       Outdoor Supply Store   0.0
4       Other Great Outdoors   0.0


----Alderwood----
            venue  freq
0     Pizza Place   2.0
1    Skating Rink   1.0
2  Sandwich Place   1.0
3             Gym   1.0
4     Coffee Shop   1.0


----Annex----
               venue  freq
0        Pizza Place   4.0
1               Park   3.0
2        Coffee Shop   3.0
3  Indian Restaurant   2.0
4             Bistro   2.0


----Banbury-Don Mills----
              venue  freq
0       Coffee Shop   1.0
1             Trail   1.0
2  Botanical Garden   1.0
3              Park   1.0
4    Ice Cream Shop   0.0


----Bathurst Manor----
        

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [30]:
#Now let's create the new dataframe and display the top 5 venues for each neighborhood.

num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
toronto_venues_sorted = pd.DataFrame(columns=columns)
toronto_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

toronto_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0               Agincourt North                  Bank   
1  Agincourt South-Malvern West        Breakfast Spot   
2                     Alderwood           Pizza Place   
3                         Annex           Pizza Place   
4             Banbury-Don Mills      Botanical Garden   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Chinese Restaurant                Bakery              Pharmacy   
1  Latin American Restaurant                Lounge     Electronics Store   
2                Coffee Shop        Sandwich Place                   Pub   
3                       Park           Coffee Shop     Indian Restaurant   
4                      Trail           Coffee Shop                  Park   

  5th Most Common Venue  
0            Beer Store  
1           Dry Cleaner  
2          Dance Studio  
3                Bistro  
4           Yoga Studio

In [31]:
toronto_venues_sorted.shape

(140, 6)

In [32]:
#merge it with demo statistics to form a new data frame
#which will be used for our model
df_toronto_merged = pd.merge(toronto_venues_sorted, df_demo[['Neighborhood', '   Chinese', 'Latitude', 'Longitude']], how='left', left_on='Neighborhood', right_on='Neighborhood')
df_toronto_merged.head()


Neighborhood 1st Most Common Venue  \
0               Agincourt North                  Bank   
1  Agincourt South-Malvern West        Breakfast Spot   
2                     Alderwood           Pizza Place   
3                         Annex           Pizza Place   
4             Banbury-Don Mills      Botanical Garden   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Chinese Restaurant                Bakery              Pharmacy   
1  Latin American Restaurant                Lounge     Electronics Store   
2                Coffee Shop        Sandwich Place                   Pub   
3                       Park           Coffee Shop     Indian Restaurant   
4                      Trail           Coffee Shop                  Park   

  5th Most Common Venue    Chinese   Latitude  Longitude  
0            Beer Store      16565  43.808038 -79.266439  
1           Dry Cleaner       9810  43.795223 -79.260241  
2          Dance Studio         70  43.601717 -79.545232  
3                Bistro       1695  43.670338 -79.407117  
4           Yoga Studio       3535  43.734804 -79.357243

In [33]:
#rename columns
df_toronto_merged.rename(columns={'   Chinese': 'No of Chinese'}, inplace =True)
df_toronto_merged.head()

Neighborhood 1st Most Common Venue  \
0               Agincourt North                  Bank   
1  Agincourt South-Malvern West        Breakfast Spot   
2                     Alderwood           Pizza Place   
3                         Annex           Pizza Place   
4             Banbury-Don Mills      Botanical Garden   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Chinese Restaurant                Bakery              Pharmacy   
1  Latin American Restaurant                Lounge     Electronics Store   
2                Coffee Shop        Sandwich Place                   Pub   
3                       Park           Coffee Shop     Indian Restaurant   
4                      Trail           Coffee Shop                  Park   

  5th Most Common Venue No of Chinese   Latitude  Longitude  
0            Beer Store         16565  43.808038 -79.266439  
1           Dry Cleaner          9810  43.795223 -79.260241  
2          Dance Studio            70  43.601717 -79.545232  
3                Bistro          1695  43.670338 -79.407117  
4           Yoga Studio          3535  43.734804 -79.357243

In [34]:
df_toronto_merged.dtypes

Neighborhood              object
1st Most Common Venue     object
2nd Most Common Venue     object
3rd Most Common Venue     object
4th Most Common Venue     object
5th Most Common Venue     object
No of Chinese             object
Latitude                 float64
Longitude                float64
dtype: object

In [35]:
#change no of chines from object to numeric
df_toronto_merged['No of Chinese'] = pd.to_numeric(df_toronto_merged['No of Chinese'])
df_toronto_merged.dtypes

Neighborhood              object
1st Most Common Venue     object
2nd Most Common Venue     object
3rd Most Common Venue     object
4th Most Common Venue     object
5th Most Common Venue     object
No of Chinese              int64
Latitude                 float64
Longitude                float64
dtype: object

## **3. Model** ##

We will be using k-means clustering to cluster the neighborhoods into 5 clusters

In [36]:
# set number of clusters
kclusters = 5

df_clustering = df_toronto_merged[['Latitude', 'Longitude']].values

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(df_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 3, 2, 0, 2, 2, 0, 0, 2], dtype=int32)

In [37]:
#adding cluster labels back into df_toronto_merged data frame
df_toronto_merged['Cluster Labels'] = kmeans.labels_
df_toronto_merged.head()


Neighborhood 1st Most Common Venue  \
0               Agincourt North                  Bank   
1  Agincourt South-Malvern West        Breakfast Spot   
2                     Alderwood           Pizza Place   
3                         Annex           Pizza Place   
4             Banbury-Don Mills      Botanical Garden   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Chinese Restaurant                Bakery              Pharmacy   
1  Latin American Restaurant                Lounge     Electronics Store   
2                Coffee Shop        Sandwich Place                   Pub   
3                       Park           Coffee Shop     Indian Restaurant   
4                      Trail           Coffee Shop                  Park   

  5th Most Common Venue  No of Chinese   Latitude  Longitude  Cluster Labels  
0            Beer Store          16565  43.808038 -79.266439               4  
1           Dry Cleaner           9810  43.795223 -79.260241               4  
2          Dance Studio             70  43.601717 -79.545232               3  
3                Bistro           1695  43.670338 -79.407117               2  
4           Yoga Studio           3535  43.734804 -79.357243               0

### **3.1. Visualising the Model** ###

In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df_toronto_merged['Latitude'], df_toronto_merged['Longitude'], df_toronto_merged['Neighborhood'], df_toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [50]:
df_toronto_merged.groupby('Cluster Labels').count()

Neighborhood  1st Most Common Venue  2nd Most Common Venue  \
Cluster Labels                                                               
0                         24                     24                     24   
1                          1                      1                      1   
2                         71                     71                     71   
3                         29                     29                     29   
4                         15                     15                     15   

                3rd Most Common Venue  4th Most Common Venue  \
Cluster Labels                                                 
0                                  24                     24   
1                                   1                      1   
2                                  71                     71   
3                                  29                     29   
4                                  15                     15   

                5th Most Common Venue  No of Chinese  Latitude  Longitude  
Cluster Labels                                                             
0                                  24             24        24         24  
1                                   1              1         1          1  
2                                  71             71        71         71  
3                                  29             29        29         29  
4                                  15             15        15         15

### **3.2. Analyzing Cluster 1** ###

In [52]:
#first cluster
df_cluster1 = df_toronto_merged.loc[df_toronto_merged['Cluster Labels'] == 0]
df_cluster1.groupby('Neighborhood')
df_cluster1.sort_values(by = 'No of Chinese', ascending = False)


Neighborhood       1st Most Common Venue  \
76                 Milliken         Japanese Restaurant   
112                 Steeles                  Playground   
65               L'Amoreaux          Chinese Restaurant   
114  Tam O'Shanter-Sullivan          Chinese Restaurant   
52        Hillcrest Village                    Pharmacy   
30       Don Valley Village              Sandwich Place   
98            Pleasant View        Fast Food Restaurant   
7           Bayview Village                        Bank   
8     Bayview Woods-Steeles                       Trail   
4         Banbury-Don Mills            Botanical Garden   
31              Dorset Park               Bowling Alley   
48               Henry Farm                Tennis Court   
42          Flemingdon Park        Fast Food Restaurant   
128        Wexford/Maryvale         Arts & Crafts Store   
29       Danforth-East York                 Coffee Shop   
57                  Ionview  Construction & Landscaping   
115           Taylor-Massey           Convenience Store   
35        East End-Danforth                 Pizza Place   
91                 Oakridge           Convenience Store   
28                 Danforth                 Coffee Shop   
121        Victoria Village   Middle Eastern Restaurant   
90        O'Connor-Parkview                 Pizza Place   
116             The Beaches                       Beach   
118        Thorncliffe Park           Afghan Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
76        Chinese Restaurant                Bakery      Asian Restaurant   
112  Health & Beauty Service           Yoga Studio   Egyptian Restaurant   
65        Athletics & Sports           Coffee Shop         Shopping Mall   
114                    Hotel                  Park  Fast Food Restaurant   
52             Grocery Store            Restaurant                  Bank   
30               Pizza Place           Coffee Shop                  Bank   
98       Japanese Restaurant           Pizza Place                  Park   
7                Pizza Place        Clothing Store         Metro Station   
8                       Park               Dog Run           Yoga Studio   
4                       Park                 Trail           Coffee Shop   
31          Asian Restaurant  Fast Food Restaurant            Beer Store   
48                Restaurant                  Park   Dumpling Restaurant   
42       Japanese Restaurant         Movie Theater              Bus Line   
128              Pizza Place      Greek Restaurant        Breakfast Spot   
29                  Pharmacy              Bus Line         Grocery Store   
57             Deli / Bodega      Business Service                  Park   
115          Warehouse Store   American Restaurant            Steakhouse   
35                    Bistro   Egyptian Restaurant          Burger Joint   
91            Ice Cream Shop            Restaurant              Bus Stop   
28                  Pharmacy              Bus Line         Grocery Store   
121          Thai Restaurant              Bus Line   Filipino Restaurant   
90      Fast Food Restaurant          Intersection        Breakfast Spot   
116      Japanese Restaurant        Breakfast Spot              Tea Room   
118              Coffee Shop     Indian Restaurant                  Bank   

           5th Most Common Venue  No of Chinese   Latitude  Longitude  \
76                  Noodle House          16790  43.823174 -79.301763   
112           Dongbei Restaurant          16705  43.816178 -79.314538   
65                   Yoga Studio          16455  43.799003 -79.305967   
114                          Bar           9615  43.768997 -79.301849   
52             Korean Restaurant           8355  43.799664 -79.365019   
30                          Park           7360  43.792673 -79.354722   
98                    Restaurant           5840  43.787048 -79.333714   
7           Fast Food Restaurant           4765  43.76

In [53]:
print('The max, min and mean of No of Chinese in cluster 1 are {}, {} and {}.'.format(df_cluster1['No of Chinese'].max(), 
                                                                                      df_cluster1['No of Chinese'].min(),
                                                                                      df_cluster1['No of Chinese'].mean()))

The max, min and mean of No of Chinese in cluster 1 are 16790, 440 and 4724.166666666667.


In [54]:
print('The top 5 most common venues for cluster 1 are {}, {}, {}, {}, {}.'.format(df_cluster1.loc[:,"1st Most Common Venue"].mode(), 
                                                                                  df_cluster1.loc[:,"2nd Most Common Venue"].mode(),
                                                                                  df_cluster1.loc[:,"3rd Most Common Venue"].mode(),
                                                                                  df_cluster1.loc[:,"4th Most Common Venue"].mode(),
                                                                                  df_cluster1.loc[:,"5th Most Common Venue"].mode()))


The top 5 most common venues for cluster 1 are 0      Chinese Restaurant
1             Coffee Shop
2       Convenience Store
3    Fast Food Restaurant
4             Pizza Place
dtype: object, 0    Japanese Restaurant
1            Pizza Place
dtype: object, 0    Bus Line
dtype: object, 0    Bank
dtype: object, 0                     Bar
1    Fast Food Restaurant
2                    Park
3               Pet Store
dtype: object.


1st Most Common Venue in Cluster 1   |2nd Most Common Venue in Cluster 1   |3rd Most Common Venue in Cluster 1   |4th Most Common Venue in Cluster 1   |5th Most Common Venue in Cluster 1   |
:---: |:---: |:---: |:---: |:---: |
Italian Restaurant   | Chinese Restaurant   |Mobile Phone Shop   |Pharmacy   |Cafe   |


- Max number of Chinese: 14860
- Min number of Chinese: 75
- Mean number of Chinese: 1447

In [58]:
df_cluster1.groupby('Neighborhood').count()
df_cluster1

Neighborhood       1st Most Common Venue  \
4         Banbury-Don Mills            Botanical Garden   
7           Bayview Village                        Bank   
8     Bayview Woods-Steeles                       Trail   
28                 Danforth                 Coffee Shop   
29       Danforth-East York                 Coffee Shop   
30       Don Valley Village              Sandwich Place   
31              Dorset Park               Bowling Alley   
35        East End-Danforth                 Pizza Place   
42          Flemingdon Park        Fast Food Restaurant   
48               Henry Farm                Tennis Court   
52        Hillcrest Village                    Pharmacy   
57                  Ionview  Construction & Landscaping   
65               L'Amoreaux          Chinese Restaurant   
76                 Milliken         Japanese Restaurant   
90        O'Connor-Parkview                 Pizza Place   
91                 Oakridge           Convenience Store   
98            Pleasant View        Fast Food Restaurant   
112                 Steeles                  Playground   
114  Tam O'Shanter-Sullivan          Chinese Restaurant   
115           Taylor-Massey           Convenience Store   
116             The Beaches                       Beach   
118        Thorncliffe Park           Afghan Restaurant   
121        Victoria Village   Middle Eastern Restaurant   
128        Wexford/Maryvale         Arts & Crafts Store   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                       Park                 Trail           Coffee Shop   
7                Pizza Place        Clothing Store         Metro Station   
8                       Park               Dog Run           Yoga Studio   
28                  Pharmacy              Bus Line         Grocery Store   
29                  Pharmacy              Bus Line         Grocery Store   
30               Pizza Place           Coffee Shop                  Bank   
31          Asian Restaurant  Fast Food Restaurant            Beer Store   
35                    Bistro   Egyptian Restaurant          Burger Joint   
42       Japanese Restaurant         Movie Theater              Bus Line   
48                Restaurant                  Park   Dumpling Restaurant   
52             Grocery Store            Restaurant                  Bank   
57             Deli / Bodega      Business Service                  Park   
65        Athletics & Sports           Coffee Shop         Shopping Mall   
76        Chinese Restaurant                Bakery      Asian Restaurant   
90      Fast Food Restaurant          Intersection        Breakfast Spot   
91            Ice Cream Shop            Restaurant              Bus Stop   
98       Japanese Restaurant           Pizza Place                  Park   
112  Health & Beauty Service           Yoga Studio   Egyptian Restaurant   
114                    Hotel                  Park  Fast Food Restaurant   
115          Warehouse Store   American Restaurant            Steakhouse   
116      Japanese Restaurant        Breakfast Spot              Tea Room   
118              Coffee Shop     Indian Restaurant                  Bank   
121          Thai Restaurant              Bus Line   Filipino Restaurant   
128              Pizza Place      Greek Restaurant        Breakfast Spot   

           5th Most Common Venue  No of Chinese   Latitude  Longitude  \
4                   Intersection           3535  43.734804 -79.357243   
7           Fast Food Restaurant           4765  43.769197 -79.376662   
8    Eastern European Restaurant           4420  43.798127 -79.382973   
28                     Pet Store            915  43.686433 -79.300355   
29                     Pet Store           1560  43.686433 -79.300355   
30                          Park           7360  43.792673 -79.354722   
31                   Gaming Cafe           3365  43.752847 -79.282067   
35                           Bar           1230  43.66

In [68]:
#find out the neighborhoods in cluster 1
df_cluster1['Neighborhood'].unique()

array(['Banbury-Don Mills', 'Bayview Village', 'Bayview Woods-Steeles',
       'Danforth', 'Danforth-East York', 'Don Valley Village',
       'Dorset Park', 'East End-Danforth', 'Flemingdon Park',
       'Henry Farm', 'Hillcrest Village', 'Ionview', "L'Amoreaux",
       'Milliken', "O'Connor-Parkview", 'Oakridge', 'Pleasant View',
       'Steeles', "Tam O'Shanter-Sullivan", 'Taylor-Massey',
       'The Beaches', 'Thorncliffe Park', 'Victoria Village',
       'Wexford/Maryvale'], dtype=object)

### **3.3. Analyzing Cluster 2** ###

In [60]:
#second cluster
df_cluster2 = df_toronto_merged.loc[df_toronto_merged['Cluster Labels'] == 1]
df_cluster2.groupby('Neighborhood')
df_cluster2.sort_values(by = 'No of Chinese', ascending = False)

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
111  St.Andrew-Windfields         Shopping Mall         Grocery Store   

      3rd Most Common Venue    4th Most Common Venue 5th Most Common Venue  \
111  Furniture / Home Store  Fruit & Vegetable Store              Pharmacy   

     No of Chinese   Latitude  Longitude  Cluster Labels  
111           3065  18.025086  -76.81517               1

As there is only 1 neighborhood under Cluster 2, we will not by performing the statistical analysis of number of chinese and the mode of each most common venue.

1st Most Common Venue in Cluster 2   |2nd Most Common Venue in Cluster 2   |3rd Most Common Venue in Cluster 2   |4th Most Common Venue in Cluster 2   |5th Most Common Venue in Cluster 2   |
:---: |:---: |:---: |:---: |:---: |
Grocery Store   | Shopping Mall   |Japanese Restaurant   |Furniture/Home Store   |Print Shop   |


- Number of Chinese: 3065


In [61]:
df_cluster2.groupby('Neighborhood').count()
df_cluster2

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
111  St.Andrew-Windfields         Shopping Mall         Grocery Store   

      3rd Most Common Venue    4th Most Common Venue 5th Most Common Venue  \
111  Furniture / Home Store  Fruit & Vegetable Store              Pharmacy   

     No of Chinese   Latitude  Longitude  Cluster Labels  
111           3065  18.025086  -76.81517               1

In [69]:
#find out the neighborhoods in cluster 2
df_cluster2['Neighborhood'].unique()

array(['St.Andrew-Windfields'], dtype=object)

### **3.4. Analyzing Cluster 3** ###

In [62]:
#third cluster
df_cluster3 = df_toronto_merged.loc[df_toronto_merged['Cluster Labels'] == 2]
df_cluster3.groupby('Neighborhood')
df_cluster3.sort_values(by = 'No of Chinese', ascending = False)

Neighborhood   1st Most Common Venue  \
129                    Willowdale East             Coffee Shop   
110                    South Riverdale      Chinese Restaurant   
62                Kensington-Chinatown                    Café   
85                    Newtonbrook East             Coffee Shop   
119                  Trinity-Bellwoods            Cocktail Bar   
95                   Parkwoods-Donalda      Italian Restaurant   
6                  Bay Street Corridor        Sushi Restaurant   
46                   Greenwood-Coxwell      Italian Restaurant   
32   Dovercourt-Wallace Emerson-Juncti      Italian Restaurant   
122  Waterfront Communities-The Island      Italian Restaurant   
23               Church-Yonge Corridor             Coffee Shop   
88                     North Riverdale      Chinese Restaurant   
24                 Clairlea-Birchmount      Italian Restaurant   
3                                Annex             Pizza Place   
130                    Willowdale West             Coffee Shop   
100                        Regent Park             Coffee Shop   
71                     Little Portugal                     Bar   
94             Palmerston-Little Italy                     Bar   
34                      Dufferin Grove                     Bar   
89                 North St.James Town             Coffee Shop   
86                    Newtonbrook West             Coffee Shop   
87                             Niagara             Yoga Studio   
12               Birchcliffe-Cliffside      Italian Restaurant   
14                         Blake-Jones      Italian Restaurant   
67                    Lansing-Westgate                     Spa   
103                       Roncesvalles                    Café   
134                   Woodbine-Lumsden      Italian Restaurant   
83                 Mount Pleasant West             Coffee Shop   
135                           Wychwood      Italian Restaurant   
16   Bridle Path-Sunnybrook-York Mills      Italian Restaurant   
40      Eringate-Centennial-West Deane      Italian Restaurant   
33                Downsview-Roding-CFB      Italian Restaurant   
104                Rosedale-Moore Park              Playground   
125                Westminster-Branson      Italian Restaurant   
18                 Brookhaven-Amesbury      Italian Restaurant   
120                         University             Coffee Shop   
124             West Humber-Clairville      Italian Restaurant   
139                 Yorkdale-Glen Park      Italian Restaurant   
92                     Oakwood Village             Pizza Place   
133                  Woodbine Corridor      Italian Restaurant   
79                           Moss Park             Coffee Shop   
68                 Lawrence Park North        Sushi Restaurant   
17                     Broadview North          Discount Store   
9                 Bedford Park-Nortown      Italian Restaurant   
81   Mount Olive-Silverstone-Jamestown      Italian Restaurant   
82                 Mount Pleasant East             Coffee Shop   
131   Willowridge-Martingrove-Richview      Italian Restaurant   
93                       Old East York             Coffee Shop   
19     Cabbagetown-South St.James Town      Italian Restaurant   
102                  Rockcliffe-Smythe      Italian Restaurant   
99                  Princess-Rosethorn      Italian Restaurant   
15                Briar Hill-Belgravia      Italian Restaurant   
63       Kingsview Village-The Westway      Italian Restaurant   
39                 Englemount-Lawrence              Playground   
136                     Yonge-Eglinton             Coffee Shop   
127                 Weston-Pellam Park      Italian Restaurant   
70                  Leaside-Bennington                    Pool   
5                       Bathurst Manor              Playground   
25                        Clanton Park    Other Great Outdoors   
55                          Humbermede      Italian Restaurant   


In [111]:
print('The max, min and mean of No of Chinese in cluster 3 are {}, {} and {}.'.format(df_cluster3['No of Chinese'].max(), 
                                                                                      df_cluster3['No of Chinese'].min(),
                                                                                      df_cluster3['No of Chinese'].mean()))

The max, min and mean of No of Chinese in cluster 3 are 4620, 320 and 1465.4545454545455.


In [112]:
print('The top 5 most common venues for cluster 3 are {}, {}, {}, {}, {}.'.format(df_cluster3.loc[:,"1st Most Common Venue"].mode(), 
                                                                                  df_cluster3.loc[:,"2nd Most Common Venue"].mode(),
                                                                                  df_cluster3.loc[:,"3rd Most Common Venue"].mode(),
                                                                                  df_cluster3.loc[:,"4th Most Common Venue"].mode(),
                                                                                  df_cluster3.loc[:,"5th Most Common Venue"].mode()))


The top 5 most common venues for cluster 3 are 0    Fast Food Restaurant
1          Ice Cream Shop
2             Pizza Place
dtype: object, 0                     Bank
1                      Bar
2           Baseball Field
3           Breakfast Spot
4              Coffee Shop
5     Fast Food Restaurant
6                     Park
7              Pizza Place
8                      Pub
9               Restaurant
10             Yoga Studio
dtype: object, 0    Discount Store
dtype: object, 0                      Beer Store
1                 Bubble Tea Shop
2                    Burger Joint
3                  Discount Store
4                         Dog Run
5              Dongbei Restaurant
6     Eastern European Restaurant
7            Fast Food Restaurant
8               Indian Restaurant
9                      Restaurant
10                    Yoga Studio
dtype: object, 0                Department Store
1              Dongbei Restaurant
2                      Donut Shop
3     Eastern European

1st Most Common Venue in Cluster 3   |2nd Most Common Venue in Cluster 3   |3rd Most Common Venue in Cluster 3   |4th Most Common Venue in Cluster 3   |5th Most Common Venue in Cluster 3   |
:---: |:---: |:---: |:---: |:---: |
Fast Food Restaurant   | Bank   |Discount Store   |Beer Store   |Department Store   |


- Max number of Chinese: 4620
- Min number of Chinese: 320
- Mean number of Chinese: 1465

In [63]:
df_cluster3.groupby('Neighborhood').count()
df_cluster3

Neighborhood   1st Most Common Venue  \
3                                Annex             Pizza Place   
5                       Bathurst Manor              Playground   
6                  Bay Street Corridor        Sushi Restaurant   
9                 Bedford Park-Nortown      Italian Restaurant   
10             Beechborough-Greenbrook      Italian Restaurant   
12               Birchcliffe-Cliffside      Italian Restaurant   
14                         Blake-Jones      Italian Restaurant   
15                Briar Hill-Belgravia      Italian Restaurant   
16   Bridle Path-Sunnybrook-York Mills      Italian Restaurant   
17                     Broadview North          Discount Store   
18                 Brookhaven-Amesbury      Italian Restaurant   
19     Cabbagetown-South St.James Town      Italian Restaurant   
20                  Caledonia-Fairbank  Furniture / Home Store   
21                           Casa Loma          Sandwich Place   
23               Church-Yonge Corridor             Coffee Shop   
24                 Clairlea-Birchmount      Italian Restaurant   
25                        Clanton Park    Other Great Outdoors   
27              Corso Italia-Davenport      Italian Restaurant   
32   Dovercourt-Wallace Emerson-Juncti      Italian Restaurant   
33                Downsview-Roding-CFB      Italian Restaurant   
34                      Dufferin Grove                     Bar   
39                 Englemount-Lawrence              Playground   
40      Eringate-Centennial-West Deane      Italian Restaurant   
43                   Forest Hill North              Playground   
44                   Forest Hill South              Playground   
46                   Greenwood-Coxwell      Italian Restaurant   
55                          Humbermede      Italian Restaurant   
56                  Humewood-Cedarvale      Italian Restaurant   
62                Kensington-Chinatown                    Café   
63       Kingsview Village-The Westway      Italian Restaurant   
67                    Lansing-Westgate                     Spa   
68                 Lawrence Park North        Sushi Restaurant   
69                 Lawrence Park South        Sushi Restaurant   
70                  Leaside-Bennington                    Pool   
71                     Little Portugal                     Bar   
79                           Moss Park             Coffee Shop   
81   Mount Olive-Silverstone-Jamestown      Italian Restaurant   
82                 Mount Pleasant East             Coffee Shop   
83                 Mount Pleasant West             Coffee Shop   
85                    Newtonbrook East             Coffee Shop   
86                    Newtonbrook West             Coffee Shop   
87                             Niagara             Yoga Studio   
88                     North Riverdale      Chinese Restaurant   
89                 North St.James Town             Coffee Shop   
92                     Oakwood Village             Pizza Place   
93                       Old East York             Coffee Shop   
94             Palmerston-Little Italy                     Bar   
95                   Parkwoods-Donalda      Italian Restaurant   
97            Playter Estates-Danforth      Italian Restaurant   
99                  Princess-Rosethorn      Italian Restaurant   
100                        Regent Park             Coffee Shop   
102                  Rockcliffe-Smythe      Italian Restaurant   
103                       Roncesvalles                    Café   
104                Rosedale-Moore Park              Playground   
110                    South Riverdale      Chinese Restaurant   
117       Thistletown-Beaumond Heights      Italian Restaurant   
119                  Trinity-Bellwoods            Cocktail Bar   
120                         University             Coffee Shop   
122  Waterfront Communities-The Island      Italian Restaurant   
124             West Humber-Clairville      Italian Restaurant   


In [70]:
#find out the neighborhoods in cluster 3
df_cluster3['Neighborhood'].unique()

array(['Annex', 'Bathurst Manor', 'Bay Street Corridor',
       'Bedford Park-Nortown', 'Beechborough-Greenbrook',
       'Birchcliffe-Cliffside', 'Blake-Jones', 'Briar Hill-Belgravia',
       'Bridle Path-Sunnybrook-York Mills', 'Broadview North',
       'Brookhaven-Amesbury', 'Cabbagetown-South St.James Town',
       'Caledonia-Fairbank', 'Casa Loma', 'Church-Yonge Corridor',
       'Clairlea-Birchmount', 'Clanton Park', 'Corso Italia-Davenport',
       'Dovercourt-Wallace Emerson-Juncti', 'Downsview-Roding-CFB',
       'Dufferin Grove', 'Englemount-Lawrence',
       'Eringate-Centennial-West Deane', 'Forest Hill North',
       'Forest Hill South', 'Greenwood-Coxwell', 'Humbermede',
       'Humewood-Cedarvale', 'Kensington-Chinatown',
       'Kingsview Village-The Westway', 'Lansing-Westgate',
       'Lawrence Park North', 'Lawrence Park South', 'Leaside-Bennington',
       'Little Portugal', 'Moss Park',
       'Mount Olive-Silverstone-Jamestown', 'Mount Pleasant East',
       'Moun

### **3.5. Analyzing Cluster 4** ###

In [64]:
#fourth cluster
df_cluster4 = df_toronto_merged.loc[df_toronto_merged['Cluster Labels'] == 3]
df_cluster4.groupby('Neighborhood')
df_cluster4.sort_values(by = 'No of Chinese', ascending = False)

Neighborhood       1st Most Common Venue  \
138       York University Heights                 Pizza Place   
45         Glenfield-Jane Heights        Fast Food Restaurant   
109                South Parkdale                        Park   
13                    Black Creek  Construction & Landscaping   
58     Islington-City Centre West                 Pizza Place   
77                         Mimico                      Bakery   
49                High Park North                        Park   
50              High Park-Swansea                        Park   
59                  Junction Area          Italian Restaurant   
41            Etobicoke West Mall                       Hotel   
113           Stonegate-Queensway         Rental Car Location   
72                    Long Branch          Light Rail Station   
75                  Markland Wood              Baseball Field   
36       Edenbridge-Humber Valley                   BBQ Joint   
106  Runnymede-Bloor West Village                        Café   
80                   Mount Dennis                 Pizza Place   
64                 Kingsway South                 Coffee Shop   
96           Pelmo Park-Humberlea       Outdoors & Recreation   
126                        Weston               Train Station   
60       Keelesdale-Eglinton West           Convenience Store   
38               Elms-Old Rexdale                 Coffee Shop   
74                     Maple Leaf                      Bakery   
84                    New Toronto          Mexican Restaurant   
54                  Humber Summit  Construction & Landscaping   
101               Rexdale-Kipling            Department Store   
66             Lambton Baby Point                 Coffee Shop   
107                        Rustic        Fast Food Restaurant   
2                       Alderwood                 Pizza Place   
53       Humber Heights-Westmount                     Dog Run   

         2nd Most Common Venue   3rd Most Common Venue  \
138       Fast Food Restaurant          Discount Store   
45               Grocery Store           Shopping Mall   
109         Light Rail Station    Gym / Fitness Center   
13           Food & Drink Shop             Coffee Shop   
58           Fish & Chips Shop                    Park   
77                         Bar     American Restaurant   
49        Gym / Fitness Center       Convenience Store   
50          Light Rail Station    Gym / Fitness Center   
59                 Coffee Shop                    Café   
41               Grocery Store             Coffee Shop   
113                Coffee Shop              Restaurant   
72   South American Restaurant            Burger Joint   
75                 Golf Course               Piano Bar   
36           Indian Restaurant                  Garden   
106                Coffee Shop                  Bakery   
80                 Coffee Shop  Furniture / Home Store   
64                Dessert Shop      Italian Restaurant   
96                        Park             Yoga Studio   
126                Coffee Shop                Pharmacy   
60              Discount Store             Coffee Shop   
38           Convenience Store     Arts & Crafts Store   
74                 Yoga Studio      Dongbei Restaurant   
84                 Coffee Shop                    Café   
54                        Park              Restaurant   
101                Pizza Place    Fast Food Restaurant   
66                    Dive Bar           Grocery Store   
107                Gas Station    Caribbean Restaurant   
2                  Coffee Shop            Dance Studio   
53                        Park             Yoga Studio   

           4th Most Common Venue        5th Most Common Venue  No of Chinese  \
138           Falafel Restaurant                  Coffee Shop           3160   
45                Discount Store                  Pizza Place           2000   
109                         Lake                        Trail           1285   
13       

In [114]:
print('The max, min and mean of No of Chinese in cluster 4 are {}, {} and {}.'.format(df_cluster4['No of Chinese'].max(), 
                                                                                      df_cluster4['No of Chinese'].min(),
                                                                                      df_cluster4['No of Chinese'].mean()))

The max, min and mean of No of Chinese in cluster 4 are 16790, 595 and 5323.8.


In [115]:
print('The top 5 most common venues for cluster 4 are {}, {}, {}, {}, {}.'.format(df_cluster4.loc[:,"1st Most Common Venue"].mode(), 
                                                                                  df_cluster4.loc[:,"2nd Most Common Venue"].mode(),
                                                                                  df_cluster4.loc[:,"3rd Most Common Venue"].mode(),
                                                                                  df_cluster4.loc[:,"4th Most Common Venue"].mode(),
                                                                                  df_cluster4.loc[:,"5th Most Common Venue"].mode()))


The top 5 most common venues for cluster 4 are 0    Fast Food Restaurant
dtype: object, 0    Coffee Shop
dtype: object, 0    Bus Line
dtype: object, 0    Park
dtype: object, 0    Chinese Restaurant
1                  Park
2             Pet Store
3            Restaurant
4           Yoga Studio
dtype: object.


1st Most Common Venue in Cluster 4   |2nd Most Common Venue in Cluster 4   |3rd Most Common Venue in Cluster 4   |4th Most Common Venue in Cluster 4   |5th Most Common Venue in Cluster 4   |
:---: |:---: |:---: |:---: |:---: |
Fast Food Restaurant   | Coffee Shop   |Bus Line   |Park   |Chinese   |


- Max number of Chinese: 16790
- Min number of Chinese: 595
- Mean number of Chinese: 5324

In [65]:
df_cluster4.groupby('Neighborhood').count()
df_cluster4

Neighborhood       1st Most Common Venue  \
2                       Alderwood                 Pizza Place   
13                    Black Creek  Construction & Landscaping   
36       Edenbridge-Humber Valley                   BBQ Joint   
38               Elms-Old Rexdale                 Coffee Shop   
41            Etobicoke West Mall                       Hotel   
45         Glenfield-Jane Heights        Fast Food Restaurant   
49                High Park North                        Park   
50              High Park-Swansea                        Park   
53       Humber Heights-Westmount                     Dog Run   
54                  Humber Summit  Construction & Landscaping   
58     Islington-City Centre West                 Pizza Place   
59                  Junction Area          Italian Restaurant   
60       Keelesdale-Eglinton West           Convenience Store   
64                 Kingsway South                 Coffee Shop   
66             Lambton Baby Point                 Coffee Shop   
72                    Long Branch          Light Rail Station   
74                     Maple Leaf                      Bakery   
75                  Markland Wood              Baseball Field   
77                         Mimico                      Bakery   
80                   Mount Dennis                 Pizza Place   
84                    New Toronto          Mexican Restaurant   
96           Pelmo Park-Humberlea       Outdoors & Recreation   
101               Rexdale-Kipling            Department Store   
106  Runnymede-Bloor West Village                        Café   
107                        Rustic        Fast Food Restaurant   
109                South Parkdale                        Park   
113           Stonegate-Queensway         Rental Car Location   
126                        Weston               Train Station   
138       York University Heights                 Pizza Place   

         2nd Most Common Venue   3rd Most Common Venue  \
2                  Coffee Shop            Dance Studio   
13           Food & Drink Shop             Coffee Shop   
36           Indian Restaurant                  Garden   
38           Convenience Store     Arts & Crafts Store   
41               Grocery Store             Coffee Shop   
45               Grocery Store           Shopping Mall   
49        Gym / Fitness Center       Convenience Store   
50          Light Rail Station    Gym / Fitness Center   
53                        Park             Yoga Studio   
54                        Park              Restaurant   
58           Fish & Chips Shop                    Park   
59                 Coffee Shop                    Café   
60              Discount Store             Coffee Shop   
64                Dessert Shop      Italian Restaurant   
66                    Dive Bar           Grocery Store   
72   South American Restaurant            Burger Joint   
74                 Yoga Studio      Dongbei Restaurant   
75                 Golf Course               Piano Bar   
77                         Bar     American Restaurant   
80                 Coffee Shop  Furniture / Home Store   
84                 Coffee Shop                    Café   
96                        Park             Yoga Studio   
101                Pizza Place    Fast Food Restaurant   
106                Coffee Shop                  Bakery   
107                Gas Station    Caribbean Restaurant   
109         Light Rail Station    Gym / Fitness Center   
113                Coffee Shop              Restaurant   
126                Coffee Shop                Pharmacy   
138       Fast Food Restaurant          Discount Store   

           4th Most Common Venue        5th Most Common Venue  No of Chinese  \
2                            Pub                 Skating Rink             70   
13                    Playground               History Museum           1230   
36                          Park                      Dog Run            315   
38       

In [71]:
#find out the neighborhoods in cluster 4
df_cluster4['Neighborhood'].unique()

array(['Alderwood', 'Black Creek', 'Edenbridge-Humber Valley',
       'Elms-Old Rexdale', 'Etobicoke West Mall',
       'Glenfield-Jane Heights', 'High Park North', 'High Park-Swansea',
       'Humber Heights-Westmount', 'Humber Summit',
       'Islington-City Centre West', 'Junction Area',
       'Keelesdale-Eglinton West', 'Kingsway South', 'Lambton Baby Point',
       'Long Branch', 'Maple Leaf', 'Markland Wood', 'Mimico',
       'Mount Dennis', 'New Toronto', 'Pelmo Park-Humberlea',
       'Rexdale-Kipling', 'Runnymede-Bloor West Village', 'Rustic',
       'South Parkdale', 'Stonegate-Queensway', 'Weston',
       'York University Heights'], dtype=object)

### **3.6. Analyzing Cluster 5** ###

In [66]:
#fifth cluster
df_cluster5 = df_toronto_merged.loc[df_toronto_merged['Cluster Labels'] == 4]
df_cluster5.groupby('Neighborhood')
df_cluster5.sort_values(by = 'No of Chinese', ascending = False)

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Agincourt North                Bakery                  Bank   
1    Agincourt South-Malvern West        Breakfast Spot                Lounge   
132                        Woburn  Fast Food Restaurant           Coffee Shop   
73                        Malvern  Fast Food Restaurant           Pizza Place   
11                        Bendale  Fast Food Restaurant          Intersection   
105                         Rouge  Fast Food Restaurant                  Park   
61                   Kennedy Park  Fast Food Restaurant    Chinese Restaurant   
37                  Eglinton East        Ice Cream Shop            Restaurant   
51                 Highland Creek           IT Services           Yoga Studio   
78                    Morningside                  Park           Coffee Shop   
123                     West Hill           Pizza Place         Grocery Store   
26                     Cliffcrest        Ice Cream Shop           Pizza Place   
22         Centennial Scarborough     Fish & Chips Shop                   Bar   
108           Scarborough Village           Coffee Shop        Discount Store   
47                      Guildwood         Train Station        Baseball Field   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Chinese Restaurant  Fast Food Restaurant    Frozen Yogurt Shop   
1    Latin American Restaurant                   ATM     Fish & Chips Shop   
132             Discount Store                  Bank     Indian Restaurant   
73                    Pharmacy  Gym / Fitness Center                  Park   
11                     Dog Run          Optical Shop    Chinese Restaurant   
105                Yoga Studio               Dog Run        Farmers Market   
61               Grocery Store      Asian Restaurant           Yoga Studio   
37              Sandwich Place     Indian Restaurant         Train Station   
51           Electronics Store    Dongbei Restaurant            Donut Shop   
78           Convenience Store     Mobile Phone Shop           Supermarket   
123                Coffee Shop  Fast Food Restaurant            Restaurant   
26              Discount Store          Burger Joint        Hardware Store   
22                        Park   Egyptian Restaurant            Donut Shop   
108         Chinese Restaurant                   Gym                   Pub   
47            Storage Facility           Yoga Studio   Egyptian Restaurant   

     No of Chinese   Latitude  Longitude  Cluster Labels  
0            16565  43.808038 -79.266439               4  
1             9810  43.795223 -79.260241               4  
132           4620  43.759824 -79.225291               4  
73            3780  43.809196 -79.221701               4  
11            3295  43.753520 -79.255336               4  
105           2610  43.804930 -79.165837               4  
61            1525  43.724878 -79.253969               4  
37            1200  43.739465 -79.232100               4  
51             940  43.790117 -79.173334               4  
78             700  43.782601 -79.204958               4  
123            660  43.768914 -79.187291               4  
26             450  43.721939 -79.236232               4  
22             420  43.787491 -79.150768               4  
108            420  43.743742 -79.211632               4  
47             320  43.755225 -79.198229               4

In [118]:
print('The max, min and mean of No of Chinese in cluster 5 are {}, {} and {}.'.format(df_cluster5['No of Chinese'].max(), 
                                                                                      df_cluster5['No of Chinese'].min(),
                                                                                      df_cluster5['No of Chinese'].mean()))

The max, min and mean of No of Chinese in cluster 5 are 9810, 50 and 849.53125.


In [119]:
print('The top 5 most common venues for cluster 5 are {}, {}, {}, {}, {}.'.format(df_cluster5.loc[:,"1st Most Common Venue"].mode(), 
                                                                                  df_cluster5.loc[:,"2nd Most Common Venue"].mode(),
                                                                                  df_cluster5.loc[:,"3rd Most Common Venue"].mode(),
                                                                                  df_cluster5.loc[:,"4th Most Common Venue"].mode(),
                                                                                  df_cluster5.loc[:,"5th Most Common Venue"].mode()))


The top 5 most common venues for cluster 5 are 0    Coffee Shop
dtype: object, 0    Coffee Shop
dtype: object, 0        Discount Store
1    Light Rail Station
2       Thai Restaurant
3           Yoga Studio
dtype: object, 0                           Café
1                    Coffee Shop
2    Eastern European Restaurant
3                    Gas Station
4                           Park
5                 Sandwich Place
dtype: object, 0    Doner Restaurant
dtype: object.


1st Most Common Venue in Cluster 5   |2nd Most Common Venue in Cluster 5   |3rd Most Common Venue in Cluster 5   |4th Most Common Venue in Cluster 5   |5th Most Common Venue in Cluster 5   |
:---: |:---: |:---: |:---: |:---: |
Coffee Shop   | Coffee Shop   |Discount Store   |Cafe   |Doner Restaurant   |


- Max number of Chinese: 9810
- Min number of Chinese: 50
- Mean number of Chinese: 850

In [67]:
df_cluster5.groupby('Neighborhood').count()
df_cluster5

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Agincourt North                Bakery                  Bank   
1    Agincourt South-Malvern West        Breakfast Spot                Lounge   
11                        Bendale  Fast Food Restaurant          Intersection   
22         Centennial Scarborough     Fish & Chips Shop                   Bar   
26                     Cliffcrest        Ice Cream Shop           Pizza Place   
37                  Eglinton East        Ice Cream Shop            Restaurant   
47                      Guildwood         Train Station        Baseball Field   
51                 Highland Creek           IT Services           Yoga Studio   
61                   Kennedy Park  Fast Food Restaurant    Chinese Restaurant   
73                        Malvern  Fast Food Restaurant           Pizza Place   
78                    Morningside                  Park           Coffee Shop   
105                         Rouge  Fast Food Restaurant                  Park   
108           Scarborough Village           Coffee Shop        Discount Store   
123                     West Hill           Pizza Place         Grocery Store   
132                        Woburn  Fast Food Restaurant           Coffee Shop   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Chinese Restaurant  Fast Food Restaurant    Frozen Yogurt Shop   
1    Latin American Restaurant                   ATM     Fish & Chips Shop   
11                     Dog Run          Optical Shop    Chinese Restaurant   
22                        Park   Egyptian Restaurant            Donut Shop   
26              Discount Store          Burger Joint        Hardware Store   
37              Sandwich Place     Indian Restaurant         Train Station   
47            Storage Facility           Yoga Studio   Egyptian Restaurant   
51           Electronics Store    Dongbei Restaurant            Donut Shop   
61               Grocery Store      Asian Restaurant           Yoga Studio   
73                    Pharmacy  Gym / Fitness Center                  Park   
78           Convenience Store     Mobile Phone Shop           Supermarket   
105                Yoga Studio               Dog Run        Farmers Market   
108         Chinese Restaurant                   Gym                   Pub   
123                Coffee Shop  Fast Food Restaurant            Restaurant   
132             Discount Store                  Bank     Indian Restaurant   

     No of Chinese   Latitude  Longitude  Cluster Labels  
0            16565  43.808038 -79.266439               4  
1             9810  43.795223 -79.260241               4  
11            3295  43.753520 -79.255336               4  
22             420  43.787491 -79.150768               4  
26             450  43.721939 -79.236232               4  
37            1200  43.739465 -79.232100               4  
47             320  43.755225 -79.198229               4  
51             940  43.790117 -79.173334               4  
61            1525  43.724878 -79.253969               4  
73            3780  43.809196 -79.221701               4  
78             700  43.782601 -79.204958               4  
105           2610  43.804930 -79.165837               4  
108            420  43.743742 -79.211632               4  
123            660  43.768914 -79.187291               4  
132           4620  43.759824 -79.225291               4

In [72]:
#find out the neighborhoods in cluster 5
df_cluster5['Neighborhood'].unique()

array(['Agincourt North', 'Agincourt South-Malvern West', 'Bendale',
       'Centennial Scarborough', 'Cliffcrest', 'Eglinton East',
       'Guildwood', 'Highland Creek', 'Kennedy Park', 'Malvern',
       'Morningside', 'Rouge', 'Scarborough Village', 'West Hill',
       'Woburn'], dtype=object)